dump/match_infra.py

```diff
@@ -0,0 +1,287 @@
+            return False
+        return True
+    
+    def sep(self, db):
+        return SonicDBConfig.getSeparator("CONFIG_DB")
+    
+    def getKeys(self, db, table, key_pattern):
+        if table not in self.db_driver:
+            return []
+        
+        all_keys = self.db_driver[table].keys()
+        key_ptrn = key_pattern
+        key_ptrn = re.escape(key_ptrn)
+        key_ptrn = key_ptrn.replace("\\*", ".*")
+        filtered_keys = []
+        for key in all_keys:
+            if re.match(key_ptrn, key):
+                filtered_keys.append(table+self.sep(db)+key)
+        return filtered_keys
+    
+    def get(self, db, key):
+        sp = self.sep(db)
+        tokens = key.split(sp)
+        key_ptrn = tokens[-1]
+        tokens.pop()
+        table = sp.join(tokens)
+        if table in self.db_driver and key_ptrn in self.db_driver[table]:
+            return self.db_driver[table][key_ptrn]
+        return {}
+            
+    def hget(self, db, key, field):
+        sp = self.sep(db)
+        tokens = key.split(sp)
+        key_ptrn = tokens[-1]
+        tokens.pop()
+        table = sp.join(tokens)
+        if table in self.db_driver and key_ptrn in self.db_driver[table] and field in self.db_driver[table][key_ptrn]:
+            return self.db_driver[table][key_ptrn][field]
+        return ""
+                        
+class MatchEngine:
+    
+    # Given a request obj, find its match in the redis
+    def fetch(self, req):
+        verbose_print(str(req))
+        template = self.__ret_template()
+        template['error']  = self.__validate_request(req)
+        if template['error']:
+            return self.__return_error(template)
+        
+        src = None
+        d_src = ""
+        if req.db:
+            d_src = req.db
+            src = RedisSource()
+        else:
+            d_src = req.file
+            src = JsonSource()
+        
+        if not src.connect(d_src, req.ns):
+            template['error']  = error_dict["CONN_ERR"]
+            return self.__return_error(template)
+        verbose_print("MatchRequest Checks Passed")
+        all_matched_keys = src.getKeys(req.db, req.table, req.key_pattern)
+        if not all_matched_keys or len(all_matched_keys) == 0:
+            template['error'] = error_dict["INV_PTTRN"]
+            return self.__return_error(template)
+        
+        filtered_keys = self.__filter_out_keys(src, req, all_matched_keys)
+        verbose_print("Filtered Keys:" + str(filtered_keys))
+        if not filtered_keys:
+            template['error'] = error_dict["NO_ENTRIES"]
+            return self.__return_error(template)
+        return self.__fill(src, req, filtered_keys)
+    
+    def __ret_template(self):
+        return {"error" : "", "keys" : [], "return_values" : {}}
+    
+    
+    def __return_error(self, template):
+        verbose_print("MatchEngine: \n" + template['error'])
+        return template
+    
+    def __validate_request(self, req):
+        
```

Comment:
Can the validation be done in MatchRequest object constructor so that no one from callers will receive an invalid MatchRequest

Reply:
Done